File: solutions/2026-03-22/javascript/2026-03-22-determine-whether-matrix-can-be-obtained-by-rotation.js
// Summary: Checks if a binary matrix 'mat' can be transformed into another binary matrix 'target' by rotating 'mat' 0, 90, 180, or 270 degrees clockwise.
// Link: https://leetcode.com/problems/determine-whether-matrix-can-be-obtained-by-rotation/
// Approach:
// The problem asks us to check if 'mat' can become 'target' after any number of 90-degree clockwise rotations.
// Since there are only 4 possible rotations (0, 90, 180, 270 degrees), we can simulate each rotation and compare the rotated matrix with 'target'.
// If any of the rotated matrices match 'target', we return true. If after checking all 4 rotations, none match, we return false.
//
// To rotate a matrix 90 degrees clockwise:
// The element at `mat[i][j]` moves to `rotated_mat[j][n - 1 - i]`, where `n` is the dimension of the matrix.
// We'll implement a helper function `rotateMatrix` that takes a matrix and returns its 90-degree clockwise rotated version.
//
// The main function `findRotation` will:
// 1. Check if the original `mat` is equal to `target`. If yes, return `true`.
// 2. Rotate `mat` 90 degrees clockwise to get `rotatedMat90`. Check if `rotatedMat90` equals `target`. If yes, return `true`.
// 3. Rotate `rotatedMat90` 90 degrees clockwise to get `rotatedMat180`. Check if `rotatedMat180` equals `target`. If yes, return `true`.
// 4. Rotate `rotatedMat180` 90 degrees clockwise to get `rotatedMat270`. Check if `rotatedMat270` equals `target`. If yes, return `true`.
// 5. If none of the above conditions are met, return `false`.
//
// A helper function `areMatricesEqual` will compare two matrices element by element.
//
// Time Complexity:
// Let N be the dimension of the square matrices (N x N).
// - Rotating a matrix takes O(N^2) time.
// - Comparing two matrices takes O(N^2) time.
// - We perform at most 4 rotations and 4 comparisons.
// Therefore, the total time complexity is O(4 * N^2) which simplifies to O(N^2).
//
// Space Complexity:
// - Each rotation creates a new matrix of size N x N.
// - The `areMatricesEqual` function uses O(1) extra space.
// - We store at most 3 rotated matrices at any given time (current, rotated 90, rotated 180, rotated 270).
// Therefore, the space complexity is O(N^2) due to the creation of rotated matrices.

/**
 * @param {number[][]} mat
 * @param {number[][]} target
 * @return {boolean}
 */
var findRotation = function(mat, target) {
    const n = mat.length;

    // Helper function to check if two matrices are equal
    const areMatricesEqual = (matrix1, matrix2) => {
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                if (matrix1[i][j] !== matrix2[i][j]) {
                    return false; // Matrices are not equal if any element mismatches
                }
            }
        }
        return true; // All elements matched, matrices are equal
    };

    // Helper function to rotate a matrix 90 degrees clockwise
    const rotateMatrix = (matrix) => {
        const rotated = Array(n).fill(0).map(() => Array(n).fill(0)); // Initialize a new n x n matrix
        for (let i = 0; i < n; i++) {
            for (let j = 0; j < n; j++) {
                // The element at original[i][j] moves to rotated[j][n - 1 - i]
                rotated[j][n - 1 - i] = matrix[i][j];
            }
        }
        return rotated; // Return the newly created rotated matrix
    };

    let currentMat = mat; // Start with the original matrix

    // Check for 0-degree rotation (original matrix)
    if (areMatricesEqual(currentMat, target)) {
        return true;
    }

    // Check for 90-degree rotation
    currentMat = rotateMatrix(currentMat);
    if (areMatricesEqual(currentMat, target)) {
        return true;
    }

    // Check for 180-degree rotation
    currentMat = rotateMatrix(currentMat);
    if (areMatricesEqual(currentMat, target)) {
        return true;
    }

    // Check for 270-degree rotation
    currentMat = rotateMatrix(currentMat);
    if (areMatricesEqual(currentMat, target)) {
        return true;
    }

    // If none of the rotations match the target, return false
    return false;
};

File: solutions/2026-03-22/java/2026-03-22-determine-whether-matrix-can-be-obtained-by-rotation.java
// Problem: Determine Whether Matrix Can Be Obtained By Rotation
// Link: https://leetcode.com/problems/determine-whether-matrix-can-be-obtained-by-rotation/
//
// Approach:
// The problem asks if a given matrix `mat` can be transformed into another matrix `target`
// by rotating `mat` by 0, 90, 180, or 270 degrees clockwise.
// We can simulate these rotations and check if any of the rotated matrices match `target`.
//
// There are four possible orientations to check:
// 1. Original matrix (0-degree rotation).
// 2. 90-degree clockwise rotation.
// 3. 180-degree clockwise rotation (equivalent to two 90-degree rotations).
// 4. 270-degree clockwise rotation (equivalent to three 90-degree rotations).
//
// To perform a 90-degree clockwise rotation of an n x n matrix:
// The element at `mat[i][j]` moves to `rotated_mat[j][n - 1 - i]`.
//
// We can implement a helper function `rotate(matrix)` that returns a new matrix
// representing the 90-degree clockwise rotation of the input matrix.
//
// Then, we check the following:
// - Is `mat` equal to `target`?
// - Rotate `mat` once and check if it equals `target`.
// - Rotate `mat` twice and check if it equals `target`.
// - Rotate `mat` thrice and check if it equals `target`.
//
// If any of these checks return true, we return true. Otherwise, after checking all
// four orientations, we return false.
//
// The comparison of two matrices `mat1` and `mat2` can be done by iterating through
// all elements and checking if `mat1[i][j] == mat2[i][j]` for all `i` and `j`.
//
// Time Complexity:
// Let `n` be the dimension of the square matrices.
// - Each rotation takes O(n^2) time to create a new matrix.
// - Comparing two matrices takes O(n^2) time.
// - We perform at most 4 rotations and 4 comparisons.
// - Therefore, the total time complexity is O(4 * n^2) = O(n^2).
//
// Space Complexity:
// - Each rotation creates a new n x n matrix, which takes O(n^2) space.
// - Since we do this at most 4 times (but we can optimize by reusing space or just storing one rotated matrix at a time),
//   the dominant space complexity comes from storing the rotated matrix.
// - Thus, the space complexity is O(n^2).
class Solution {
    // Function to check if two matrices are equal.
    private boolean areMatricesEqual(int[][] mat1, int[][] mat2) {
        int n = mat1.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (mat1[i][j] != mat2[i][j]) {
                    return false; // Matrices differ at this element
                }
            }
        }
        return true; // All elements are equal
    }

    // Function to rotate a matrix 90 degrees clockwise.
    // Returns a new matrix, does not modify the original.
    private int[][] rotate90Clockwise(int[][] matrix) {
        int n = matrix.length;
        int[][] rotatedMatrix = new int[n][n]; // Create a new matrix for the rotated version

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                // The element at matrix[i][j] moves to rotatedMatrix[j][n - 1 - i]
                rotatedMatrix[j][n - 1 - i] = matrix[i][j];
            }
        }
        return rotatedMatrix; // Return the newly created rotated matrix
    }

    // Main function to determine if target can be obtained by rotating mat.
    public boolean findRotation(int[][] mat, int[][] target) {
        // Check all four possible rotations: 0, 90, 180, 270 degrees.
        
        int[][] currentMat = mat; // Start with the original matrix

        // Check 0-degree rotation (original matrix)
        if (areMatricesEqual(currentMat, target)) {
            return true;
        }

        // Check 90-degree rotation
        currentMat = rotate90Clockwise(currentMat); // Rotate mat by 90 degrees
        if (areMatricesEqual(currentMat, target)) {
            return true;
        }

        // Check 180-degree rotation (rotate again)
        currentMat = rotate90Clockwise(currentMat); // Rotate the already 90-degree rotated matrix by 90 degrees
        if (areMatricesEqual(currentMat, target)) {
            return true;
        }

        // Check 270-degree rotation (rotate again)
        currentMat = rotate90Clockwise(currentMat); // Rotate the already 180-degree rotated matrix by 90 degrees
        if (areMatricesEqual(currentMat, target)) {
            return true;
        }

        // If none of the rotations match the target, return false.
        return false;
    }
}
